File: lessons/compiled_lang/go/src/o00.flags.go
package main

import (
	"fmt"
	"os"
)

// The "flag" package enforces its own conventions (e.g. mandatory "--"
//  prefixes, and no clean way to handle "-?") that don't match this
//  lesson's exact flag set, so os.Args is walked by hand instead.
func usage(out *os.File, scriptName string) {
	fmt.Fprintln(out)
	fmt.Fprintln(out, "Usage: "+scriptName+" [-c|-e|-l|-k|-p|-m|-t] [-h|-?]")
	fmt.Fprintln(out)
	fmt.Fprintln(out, "  -c  Coffee")
	fmt.Fprintln(out, "  -e  Espresso")
	fmt.Fprintln(out, "  -l  Latte")
	fmt.Fprintln(out, "  -k  Machiato")
	fmt.Fprintln(out, "  -p  Capucino")
	fmt.Fprintln(out, "  -m  Mocha")
	fmt.Fprintln(out, "  -t  Tea")
	fmt.Fprintln(out, "  -h  Display this help message")
	fmt.Fprintln(out, "  -?  Display this help message")
	fmt.Fprintln(out)
}

func main() {
	scriptName := os.Args[0]
	args := os.Args[1:]

	if len(args) == 0 {
		usage(os.Stderr, scriptName)
		os.Exit(1)
	}

	switch args[0] {
	case "-c":
		fmt.Println("You ordered a Coffee.")
	case "-e":
		fmt.Println("You ordered an Espresso.")
	case "-l":
		fmt.Println("You ordered a Latte.")
	case "-k":
		fmt.Println("You ordered a Machiato.")
	case "-p":
		fmt.Println("You ordered a Capucino.")
	case "-m":
		fmt.Println("You ordered a Mocha.")
	case "-t":
		fmt.Println("You ordered a Tea.")
	case "-h", "-?":
		usage(os.Stdout, scriptName)
	default:
		usage(os.Stderr, scriptName)
		os.Exit(1)
	}
}

File: lessons/compiled_lang/rust/src/o00.flags.rs
use std::env;
use std::process::exit;

fn usage(script_name: &str) -> String {
    format!("\nUsage: {script_name} [-c|-e|-l|-k|-p|-m|-t] [-h|-?]\n\n  -c  Coffee\n  -e  Espresso\n  -l  Latte\n  -k  Machiato\n  -p  Capucino\n  -m  Mocha\n  -t  Tea\n  -h  Display this help message\n  -?  Display this help message\n\n")
}

fn main() {
    // env::args() reports the whole invoked path (e.g. ".\bin\o00.flags.exe"),
    //  not a basename - same as j00.arguments.rs, and unlike a shell
    //  lesson's own `basename "$0"`, since nothing here strips it.
    let all_args: Vec<String> = env::args().collect();
    let script_name = &all_args[0];
    let args = &all_args[1..];

    // Only the first flag is ever examined - equivalent to a single pass
    //  through getopts that returns after the first recognized option.
    match args.first().map(String::as_str) {
        Some("-c") => { println!("You ordered a Coffee."); exit(0); }
        Some("-e") => { println!("You ordered an Espresso."); exit(0); }
        Some("-l") => { println!("You ordered a Latte."); exit(0); }
        Some("-k") => { println!("You ordered a Machiato."); exit(0); }
        Some("-p") => { println!("You ordered a Capucino."); exit(0); }
        Some("-m") => { println!("You ordered a Mocha."); exit(0); }
        Some("-t") => { println!("You ordered a Tea."); exit(0); }
        Some("-h") | Some("-?") => { print!("{}", usage(script_name)); exit(0); }
        _ => { eprint!("{}", usage(script_name)); exit(1); }
    }
}
